mailpoet/lib/Automation/Engine/Exceptions/Exception.php

```diff
@@ -2,12 +2,14 @@
 
 namespace MailPoet\Automation\Engine\Exceptions;
 
+use Exception as PhpException;
+use MailPoet\API\REST\Exception as RestException;
 use Throwable;
 
 /**
  * Frames all MailPoet Automation exceptions ("$e instanceof MailPoet\Automation\Exception").
  */
-abstract class Exception extends \Exception {
+abstract class Exception extends PhpException implements RestException {
```

Comment:
It seems a little bit strange on its face to have all Automation exceptions implement a `RestException` interface. Doesn't that kind of imply that we should only use these exceptions in code that's run as part of a REST request? I suppose that's probably true most of the time, unless action scheduler events aren't triggered as part of a REST request...

Anyway, I'm sure this is fine, it just struck me as a little strange when I first saw it.

Reply:
Hmm, yeah, that's a good point, it does feel a bit strange now that all of these exceptions implement `MailPoet\API\REST\Exception`. I guess it partly comes from the extraction to `MailPoet\API\REST` while I wanted to simply keep the same behavior as before.

In practice, this is quite a handy thing to do — we don't have to think about some complex structure where some exceptions implement the interface, and some don't. It's also not "extending" a REST exception, only happens to implement a REST interface, which I guess still is slightly better.

We could maybe say that the `getStatusCode()`, `getErrorCode()`, and `getErrors()` are a universal thing we want to have for all exceptions, and then maybe the interface could be just `MailPoet\Exception` while the API would just "happen to know" how to process them. The weird part still is that the "status code" happens to correspond with HTTP status codes, so yeah, this is not a great abstraction either.

I don't know what's a good alternative here now. Designing an exception hierarchy is surprisingly hard, and it almost never feels perfect. And even when we have one, then it's really hard to have all devs use it correctly. I guess, ideally, anywhere deeper than in the endpoint (or maybe controller) classes, we should use just some generic exceptions, not REST-aware or anything, but that feels quite hard to do — we'd have to either catch them in endpoints and convert to error responses or they'd still need some kind of conversion to an HTTP status.

Do you have any ideas on how to improve this? And maybe, in general, the usage of exceptions in the code? I guess we don't have to solve it here and now, but it surely is one of those things that when we postpone it, we'd never do it, so maybe we could create a ticket or a P2 post to discuss how we could improve this.

Reply:
Thanks for your thoughts on this! I'm sorry to say I don't have any great ideas for different approaches. I'll use this as an excuse:

> Designing an exception hierarchy is surprisingly hard, and it almost never feels perfect.

I think we should still create a backlog ticket for this if for no other reason than to already have one the next time this topic comes up, but it doesn't seem like a very high priority maintenance item even, especially since these exceptions are currently working just as we need them to. 

If you think it would be worth it at this point to collect other people's feedback and think a P2 would be fruitful, however, please don't let me stop you!